src/replication.c

```diff
@@ -1454,25 +1466,48 @@ redisDb *disklessLoadMakeBackups(void) {
  *
  * When instead the loading succeeded we want just to free our old backups,
  * in that case the function will do just that when 'restore' is 0. */
-void disklessLoadRestoreBackups(redisDb *backup, int restore, int empty_db_flags)
+void disklessLoadRestoreBackups(redisDb *db_backup, slotsToKeysMap *map_backup,
+                                int restore, int empty_db_flags)
 {
     if (restore) {
         /* Restore. */
         emptyDbGeneric(server.db,-1,empty_db_flags,replicationEmptyDbCallback);
```

Comment:
if we proceed this way, i think we should eliminate this call to emptyDbGeneric too.
and instead move the call to emptyDb when rdbLoadRio fails outside the `else` to be before the `if` for REPL_DISKLESS_LOAD_SWAPDB.

however, i imagine that the original intent behind utilizing emptyDbGeneric for the backup (both recovery and discard) was that we preferred to have emptyDbGeneric have some segments in `if` rather than clone the common segments to a different place as you did.
One of the reasons is that this way (cloning code elsewhere), it's easier to forget to update it when something is added to emptyDb. (unlike the current approach, which is staring you in the face).

i think that in some way, backup/recovery/discard should be a feature of db.c.
although as you realized, similar things should be done for cluster and also modules (they must get a notification to backup / restore their out of keyspace data), but maybe all of that should be put in one area of db.c rather than replication.c?
@yossigo @guybe7 please share your thoughts.